pkg/mcs/tso/server/server.go

```diff
@@ -16,98 +16,292 @@ package server
+	// The entire key is in the format of "/pd/<cluster-id>/microservice/tso/keyspace-group-XXXXX/primary" in which
+	// XXXXX is 5 digits integer with leading zeros. For now we use 0 as the default cluster id.
+	tsoKeyspaceGroupPrimaryElectionPrefix = "/pd/0/microservice/tso/keyspace-group-"
+	// defaultGRPCGracefulStopTimeout is the default timeout to wait for grpc server to gracefully stop
+	defaultGRPCGracefulStopTimeout = 5 * time.Second
+	// defaultHTTPGracefulShutdownTimeout is the default timeout to wait for http server to gracefully shutdown
+	defaultHTTPGracefulShutdownTimeout = 5 * time.Second
+)
+
 var _ bs.Server = (*Server)(nil)
+var _ tso.Member = (*member.Participant)(nil)
 
-// Server is the TSO server, and it implements bs.Server and tso.GrpcServer.
+// Server is the TSO server, and it implements bs.Server.
 type Server struct {
 	diagnosticspb.DiagnosticsServer
 
+	// Server state. 0 is not serving, 1 is serving.
+	isServing int64
 	// Server start timestamp
 	startTimestamp int64
 
-	ctx       context.Context
-	name      string
-	clusterID uint64
+	ctx              context.Context
+	serverLoopCtx    context.Context
+	serverLoopCancel func()
+	serverLoopWg     sync.WaitGroup
+
+	cfg         *Config
+	clusterID   uint64
+	rootPath    string
+	backendUrls []url.URL
+
+	// for the primary election in the TSO cluster
+	participant *member.Participant
 	// etcd client
-	client *clientv3.Client
+	etcdClient *clientv3.Client
 	// http client
-	httpClient          *http.Client
+	httpClient *http.Client
+
+	muxListener         net.Listener
+	service             *Service
 	tsoAllocatorManager *tso.AllocatorManager
 	// Store as map[string]*grpc.ClientConn
 	clientConns sync.Map
 	// Store as map[string]chan *tsoRequest
 	tsoDispatcher sync.Map
+
 	// Callback functions for different stages
 	// startCallbacks will be called after the server is started.
 	startCallbacks []func()
+	// primaryCallbacks will be called after the server becomes the primary.
+	primaryCallbacks []func(context.Context)
 }
 
-// NewServer creates a new TSO server.
-func NewServer(ctx context.Context, client *clientv3.Client, httpClient *http.Client,
-	dxs diagnosticspb.DiagnosticsServer) *Server {
-	return &Server{
-		DiagnosticsServer: dxs,
-		startTimestamp:    time.Now().Unix(),
-		ctx:               ctx,
-		name:              "TSO",
-		client:            client,
-		httpClient:        httpClient,
-	}
-}
-
-// TODO: Implement the following methods defined in bs.Server
+// Implement the following methods defined in bs.Server
 
-// Name returns the unique etcd Name for this server in etcd cluster.
+// Name returns the unique Name for this server in the TSO cluster.
 func (s *Server) Name() string {
-	return s.name
+	return s.cfg.Name
 }
 
 // Context returns the context of server.
 func (s *Server) Context() context.Context {
 	return s.ctx
 }
 
+// GetNetListener returns cmux listerner of grpc and http server
+func (s *Server) GetNetListener() net.Listener {
```

Comment:
where do we use it?

Reply:
I'll remove it. I was adding another test but it was reverted it and make it to be unused.

Reply:
removed locally.